File: pep249_dbapi/tests/integ/test_put_get_simple.py
import tempfile
from pathlib import Path

from .utils_put_get import (
    as_file_uri,
    create_temporary_stage,
    shared_test_data_dir,
    PUT_ROW_SOURCE_IDX,
    PUT_ROW_TARGET_IDX,
    PUT_ROW_SOURCE_SIZE_IDX,
    PUT_ROW_TARGET_SIZE_IDX,
    PUT_ROW_SOURCE_COMPRESSION_IDX,
    PUT_ROW_TARGET_COMPRESSION_IDX,
    PUT_ROW_STATUS_IDX,
    PUT_ROW_MESSAGE_IDX,
    GET_ROW_FILE_IDX,
    GET_ROW_SIZE_IDX,
    GET_ROW_STATUS_IDX,
    GET_ROW_MESSAGE_IDX,
    LS_ROW_NAME_IDX,
)

from ..utils import NEW_DRIVER_ONLY, OLD_DRIVER_ONLY


def test_put_select(cursor):
    stage_name = create_temporary_stage(cursor, "PYTEST_STAGE_PUT_SELECT")
    _, file_path = uncompressed_test_file()

    # Upload the file to the stage
    cursor.execute(
        f"PUT 'file://{as_file_uri(file_path)}' @{stage_name}"
    )

    # Query the staged file and verify the content
    select_sql = f"SELECT $1, $2, $3 FROM @{stage_name}"
    cursor.execute(select_sql)
    row = cursor.fetchone()
    assert row == ("1", "2", "3")


def test_put_ls(cursor):
    stage_name = create_temporary_stage(cursor, "PYTEST_STAGE_PUT_LS")
    filename, file_path = uncompressed_test_file()

    # Upload the file to the stage
    cursor.execute(
        f"PUT 'file://{as_file_uri(file_path)}' @{stage_name}"
    )

    # List stage contents and verify the gzipped file is present
    expected_filename = f"{stage_name.lower()}/{filename}.gz"
    cursor.execute(f"LS @{stage_name}")
    row = cursor.fetchone()

    assert row[LS_ROW_NAME_IDX] == expected_filename


def test_get(cursor):
    stage_name = create_temporary_stage(cursor, "PYTEST_STAGE_GET")
    filename, file_path = uncompressed_test_file()
    _, compressed_file_path = compressed_test_file()

    # Upload the file to the stage
    cursor.execute(
        f"PUT 'file://{as_file_uri(file_path)}' @{stage_name}"
    )

    with tempfile.TemporaryDirectory() as tmp:
        download_dir = Path(tmp) / "download"
        download_dir.mkdir(parents=True, exist_ok=True)

        # Download with GET into local directory
        cursor.execute(
            f"GET @{stage_name}/{filename} 'file://{as_file_uri(download_dir)}/'"
        )

        # Expect gzipped file at destination
        expected_file = download_dir / f"{filename}.gz"
        assert expected_file.exists(), f"Expected downloaded file at {expected_file}"

        # Verify content
        downloaded_content = expected_file.read_bytes()
        reference_content = compressed_file_path.read_bytes()

        if OLD_DRIVER_ONLY("BC#1"):
            # Old driver adds extra gzip headers, compressed content should differ
            assert downloaded_content != reference_content

        if NEW_DRIVER_ONLY("BC#1"):
            # New driver matches gzip content exactly
            assert downloaded_content == reference_content


def test_put_get_rowset(cursor):
    stage_name = create_temporary_stage(cursor, "PYTEST_STAGE_PUT_ROWSET")
    filename, file_path = uncompressed_test_file()

    # Upload the file to the stage
    cursor.execute(f"PUT 'file://{as_file_uri(file_path)}' @{stage_name}")

    # Verify the upload result
    row = cursor.fetchone()
    assert row[PUT_ROW_SOURCE_IDX] == filename
    assert row[PUT_ROW_TARGET_IDX] == f"{filename}.gz"
    assert row[PUT_ROW_SOURCE_SIZE_IDX] == 6

    if OLD_DRIVER_ONLY("BC#1"):
        assert row[PUT_ROW_TARGET_SIZE_IDX] == 48

    if NEW_DRIVER_ONLY("BC#1"):
        assert row[PUT_ROW_TARGET_SIZE_IDX] == 32

    assert row[PUT_ROW_SOURCE_COMPRESSION_IDX] == "NONE"
    assert row[PUT_ROW_TARGET_COMPRESSION_IDX] == "GZIP"
    assert row[PUT_ROW_STATUS_IDX] == "UPLOADED"
    assert row[PUT_ROW_MESSAGE_IDX] == ""

    with tempfile.TemporaryDirectory() as tmp:
        tmpdir_path = Path(tmp)
        # Download the file from the stage
        cursor.execute(f"GET @{stage_name}/{filename} 'file://{as_file_uri(tmpdir_path)}/'")

        # Verify the download result
        row = cursor.fetchone()

        assert row[GET_ROW_FILE_IDX] == f"{filename}.gz"

        if OLD_DRIVER_ONLY("BC#1"):
            assert row[GET_ROW_SIZE_IDX] == 42

        if NEW_DRIVER_ONLY("BC#1"):
            assert row[GET_ROW_SIZE_IDX] == 26

        assert row[GET_ROW_STATUS_IDX] == "DOWNLOADED"
        assert row[GET_ROW_MESSAGE_IDX] == ""


def uncompressed_test_file():
    return "test_data.csv", shared_test_data_dir() / "compression" / "test_data.csv"


def compressed_test_file():
    return "test_data.csv.gz", shared_test_data_dir() / "compression" / "test_data.csv.gz"

File: odbc_tests/tests/put_get_tests/test_put_get_simple.cpp
#include <algorithm>
#include <cctype>
#include <filesystem>
#include <fstream>
#include <random>
#include <string>

#include <catch2/catch_test_macros.hpp>

#include "Connection.hpp"
#include "get_data.hpp"
#include "put_get_utils.hpp"
#include "utils.hpp"

using namespace pg_utils;
namespace fs = std::filesystem;

static std::string to_lower_copy(const std::string& s) {
  std::string out = s;
  std::transform(out.begin(), out.end(), out.begin(),
                 [](unsigned char c) { return std::tolower(c); });
  return out;
}

static std::pair<std::string, fs::path> basic_test_file() {
  return {"test_data.csv", test_utils::shared_test_data_dir() / "basic" / "test_data.csv"};
}

TEST_CASE("PUT then SELECT from stage", "[put_get][odbc]") {
  Connection conn;
  const std::string stage = pg_utils::create_stage(conn, "ODBCTST_PUT_SELECT");

  // Use shared test data
  auto [filename, file] = basic_test_file();

  // Setup stage and upload file
  std::string put_sql = "PUT 'file://" + as_file_uri(file) + "' @" + stage;
  conn.execute(put_sql);

  {
    // Query the uploaded file data
    std::string select_sql = "SELECT $1, $2, $3 FROM @" + stage;
    auto stmt = conn.execute_fetch(select_sql);

    // Verify the data matches what we uploaded
    CHECK(get_data<SQL_C_CHAR>(stmt, 1) == "1");
    CHECK(get_data<SQL_C_CHAR>(stmt, 2) == "2");
    CHECK(get_data<SQL_C_CHAR>(stmt, 3) == "3");
  }
}

TEST_CASE("PUT then LS shows gz file", "[put_get][odbc]") {
  Connection conn;
  const std::string stage = pg_utils::create_stage(conn, "ODBCTST_PUT_LS");
  auto [filename, file] = basic_test_file();

  // Upload file
  std::string put_sql = "PUT 'file://" + as_file_uri(file) + "' @" + stage;
  conn.execute(put_sql);

  // Verify file was uploaded with LS command
  {
    std::string ls_sql = "LS @" + stage;
    auto stmt = conn.execute_fetch(ls_sql);
    std::string name = get_data<SQL_C_CHAR>(stmt, LS_ROW_NAME_IDX);
    std::string expected = to_lower_copy(stage) + "/" + filename + ".gz";
    CHECK(name == expected);
  }
}

TEST_CASE("GET downloads file to directory", "[put_get][odbc]") {
  Connection conn;
  const std::string stage = pg_utils::create_stage(conn, "ODBCTST_GET");
  auto [filename, file] = basic_test_file();

  // PUT file
  std::string put_sql = "PUT 'file://" + as_file_uri(file) + "' @" + stage;
  conn.execute(put_sql);

  // GET into download dir
  fs::path download_dir = fs::temp_directory_path() / (std::string("odbc_put_get_") + random_hex());
  fs::create_directories(download_dir);
  {
    std::string get_sql =
        "GET @" + stage + "/" + filename + " 'file://" + as_file_uri(download_dir) + "/'";
    auto stmt = conn.execute_fetch(get_sql);
    CHECK(get_data<SQL_C_CHAR>(stmt, GET_ROW_FILE_IDX) == filename + ".gz");
  }

  // Verify the downloaded file exists and content matches
  fs::path gz = download_dir / (filename + ".gz");
  REQUIRE(fs::exists(gz));

  // Decompress and verify content
  std::string decompressed = decompress_gzip_file(gz);
  std::ifstream ifs(file);
  std::string original_content((std::istreambuf_iterator<char>(ifs)),
                               std::istreambuf_iterator<char>());
  CHECK(decompressed == original_content);
}

// BREAKING CHANGE: Compression type is now returned in uppercase
TEST_CASE("PUT then GET returns expected rowset metadata", "[put_get][odbc]") {
  Connection conn;
  const std::string stage = pg_utils::create_stage(conn, "ODBCTST_PUT_ROWSET");
  auto [filename, file] = basic_test_file();

  {
    // Upload file
    auto stmt = conn.execute_fetch("PUT 'file://" + as_file_uri(file) + "' @" + stage);

    // Assert PUT result fields (file, gz target, sizes, compression, status, message)
    CHECK(get_data<SQL_C_CHAR>(stmt, PUT_ROW_SOURCE_IDX) == filename);
    CHECK(get_data<SQL_C_CHAR>(stmt, PUT_ROW_TARGET_IDX) == filename + ".gz");
    CHECK(get_data<SQL_C_LONG>(stmt, PUT_ROW_SOURCE_SIZE_IDX) == 6);
    CHECK(get_data<SQL_C_LONG>(stmt, PUT_ROW_TARGET_SIZE_IDX) == 32);
    compare_compression_type(get_data<SQL_C_CHAR>(stmt, PUT_ROW_SOURCE_COMPRESSION_IDX), "NONE");
    compare_compression_type(get_data<SQL_C_CHAR>(stmt, PUT_ROW_TARGET_COMPRESSION_IDX), "GZIP");
    CHECK(get_data<SQL_C_CHAR>(stmt, PUT_ROW_STATUS_IDX) == "UPLOADED");

    OLD_DRIVER_ONLY("BD#3: Encryption field is no longer included in the result") {
      CHECK(get_data<SQL_C_CHAR>(stmt, 8) == "ENCRYPTED");
      CHECK(get_data<SQL_C_CHAR>(stmt, 9) == "");
    }

    NEW_DRIVER_ONLY("BD#3: Encryption field is no longer included in the result") {
      CHECK(get_data<SQL_C_CHAR>(stmt, PUT_ROW_MESSAGE_IDX) == "");  // 8
    }
  }

  // Create directory for download
  fs::path download_dir = fs::temp_directory_path() / (std::string("odbc_put_get_") + random_hex());
  fs::create_directories(download_dir);
  {
    // Download file
    auto stmt = conn.execute_fetch("GET @" + stage + "/" + filename + " 'file://" +
                                   as_file_uri(download_dir) + "/'");

    // Assert GET result fields (file, size, status, message)
    CHECK(get_data<SQL_C_CHAR>(stmt, GET_ROW_FILE_IDX) == filename + ".gz");

    OLD_DRIVER_ONLY("BD#4: GET rowset metadata contains file size after decryption") {
      CHECK(get_data<SQL_C_LONG>(stmt, GET_ROW_SIZE_IDX) == 32);
    }

    NEW_DRIVER_ONLY("BD#4: GET rowset metadata contains file size after decryption") {
      CHECK(get_data<SQL_C_LONG>(stmt, GET_ROW_SIZE_IDX) == 26);
    }

    CHECK(get_data<SQL_C_CHAR>(stmt, GET_ROW_STATUS_IDX) == "DOWNLOADED");

    OLD_DRIVER_ONLY("BD#3: Encryption field is no longer included in the result") {
      CHECK(get_data<SQL_C_CHAR>(stmt, 4) == "DECRYPTED");
      CHECK(get_data<SQL_C_CHAR>(stmt, 5) == "");
    }

    NEW_DRIVER_ONLY("BD#3: Encryption field is no longer included in the result") {
      CHECK(get_data<SQL_C_CHAR>(stmt, GET_ROW_MESSAGE_IDX) == "");  // 4
    }
  }
}
